plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -642,52 +842,28 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
    }
 
-   void update_active_finalizers() {
-      if (!_is_savanna_active || _producers.empty())
-         return;
-
-      auto& chain  = chain_plug->chain();
-      finalizer_policy_ptr fin_policy = chain.head_active_finalizer_policy();
-      assert(fin_policy);
-      _is_producer_active_finalizer = std::ranges::any_of(fin_policy->finalizers, [&](const auto& f) {
-         return _producers.contains(to_account_name_safe(f.description));
-      });
-      _other_active_finalizers = std::ranges::any_of(fin_policy->finalizers, [&](const auto& f) {
-         return !_producers.contains(to_account_name_safe(f.description));
-      });
-      if (!_is_producer_active_finalizer || !_other_active_finalizers) {
-         if (fin_policy = chain.head_pending_finalizer_policy(); fin_policy) {
-            if (!_is_producer_active_finalizer) {
-               _is_producer_active_finalizer = std::ranges::any_of(fin_policy->finalizers, [&](const auto& f) {
-                  return _producers.contains(to_account_name_safe(f.description));
-               });
-            }
-            if (!_other_active_finalizers) {
-               _other_active_finalizers = std::ranges::any_of(fin_policy->finalizers, [&](const auto& f) {
-                  return !_producers.contains(to_account_name_safe(f.description));
-               });
-            }
-         }
-      }
+   bool implicitly_paused() const {
+      return _implicit_pause_vote_tracker.implicitly_paused() != production_pause_vote_tracker::implicit_pause::not_paused;
    }
 
-   void on_block(const signed_block_ptr& block, const block_id_type& id) {
+   void on_accepted_block(const signed_block_ptr& block, const block_id_type& id) {
       auto& chain  = chain_plug->chain();
       auto  before = _unapplied_transactions.size();
       _unapplied_transactions.clear_applied(block);
-      auto now = fc::time_point::now();
-      chain.get_mutable_subjective_billing().on_block(_log, block, now);
-      _accepted_block_time = now;
-      update_active_finalizers();
       if (before > 0) {
-         fc_dlog(_log, "Removed applied transactions before: ${before}, after: ${after}", ("before", before)("after", _unapplied_transactions.size()));
+         fc_dlog(_log, "Removed applied transactions before: ${before}, after: ${after}",
+                 ("before", before)("after", _unapplied_transactions.size()));
       }
+      auto now = fc::time_point::now();
+      chain.get_mutable_subjective_billing().on_block(_log, block, now);
    }
 
-   void on_block_header(const signed_block_ptr& block) {
+   void on_accepted_block_header(const signed_block_ptr& block) {
       if (!block->is_proper_svnn_block()) {
          if (_producers.contains(block->producer))
             _producer_watermarks.consider_new_watermark(block->producer, block->block_num(), block->timestamp);
+      } else {
+         _implicit_pause_vote_tracker.on_accepted_block_header();
```

Comment:
I think this should still first call the `producer_plugin_impl::update_active_finalizers();` function which should be returned back.

And then after that you would instead make the call:

```cpp
_implicit_pause_vote_tracker.record_received_block(fc::time_point::now());
```